File: src/runtime/storageNG/storage-key-parser.ts
/**
 * @license
 * Copyright 2019 Google LLC.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
import {StorageKey} from './storage-key.js';
import {VolatileStorageKey} from './drivers/volatile.js';
import {RamDiskStorageKey} from './drivers/ramdisk.js';
import {FirebaseStorageKey} from './drivers/firebase.js';
import {ReferenceModeStorageKey} from './reference-mode-storage-key.js';

type ParserTopLevel = (key: string) => StorageKey;
type Parser = (key: string, parse: ParserTopLevel) => StorageKey;

/**
 * Parses storage key string representations back into real StorageKey
 * instances.
 *
 * Singleton class with static methods. If you modify the default set of storage
 * keys in a test, remember to call StorageKeyParser.reset() in the tear-down
 * method.
 */
export class StorageKeyParser {
  private static parsers = StorageKeyParser.getDefaultParsers();

  private static getDefaultParsers(): Map<string, Parser> {
    return new Map<string, Parser>([
      ['volatile', VolatileStorageKey.fromString],
      ['firebase', FirebaseStorageKey.fromString],
      ['ramdisk', RamDiskStorageKey.fromString],
      ['reference-mode', ReferenceModeStorageKey.fromString]
    ]);
  }

  static parse(key: string): StorageKey {
    const match = key.match(/^((?:\w|-)+):\/\/(.*)$/);
    if (!match) {
      throw new Error('Failed to parse storage key: ' + key);
    }
    const protocol = match[1];
    const parser = StorageKeyParser.parsers.get(protocol);
    if (!parser) {
      throw new Error(`Unknown storage key protocol ${protocol} in key ${key}.`);
    }
    return parser(key, StorageKeyParser.parse);
  }

  static reset() {
    this.parsers = this.getDefaultParsers();
  }

  static addParser(protocol: string, parser: Parser) {
    if (this.parsers.has(protocol)) {
      throw new Error(`Parser for storage key protocol ${protocol} already exists.`);
    }
    this.parsers.set(protocol, parser);
  }
}

File: java/arcs/core/storage/StorageKeyParser.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.storage

/**
 * Parses storage key string representations back into real StorageKey
 * instances.
 *
 * If you modify the default set of storage keys in a test (using [addParser]), remember to call
 * [reset] in the tear-down method.
 */
interface StorageKeyManager {
  /**
   * Return a structured [StorageKey] instance for the provided raw key string. Implementations
   * may throw an exception if:
   *   * The key string has invalid structure.
   *   * The key string uses a protocol that doesn't have a registered parser.
   *   * The key doesn't match the parser's structural expectations.
   */
  fun parse(rawKeyString: String): StorageKey

  /**
   * Add a new [StorageKeyParser] to the internal list of parsers. Adding a [parser] for a
   * protocol which already exists should replace the existing implementation.
   */
  fun addParser(parser: StorageKeyParser<*>)

  /**
   * Remove all currently registered parsers, and replace them with the values provided.
   */
  fun reset(vararg initialSet: StorageKeyParser<*>)
}

/**
 * The interface for a parser of a specific protocol type.
 *
 * By convention, when implementing [StorageKey], you should also include in the companion object
 * a [protocol] field that defines the protocol for the name, and a [parser] that implements this
 * interface.
 *
 * The companion object for this interface implements a thread-safe global instance of
 * [StorageKeyManager], and that's currently what we use throughout the Arcs codebase for storage
 * key management.
 */
interface StorageKeyParser<T : StorageKey> {
  /** The protocol that this [StorageKeyParser] supports. */
  val protocol: String

  /** Returns a structured key of type [T] give the [rawKeyString]. May throw an exception if:
   *   * The [rawKeyString] has invalid structure.
   *   * The structure of the key doesn't meet the specific requirements for this [protocol].
   *
   *   Note that implementations are not guaranteed to check the protocol of the [rawKeyString]
   *   provided, so callers should verify this themselves before using the parser.
   */
  fun parse(rawKeyString: String): T

  /**
   * Expose the [DefaultStorageKeyManager] via the [StorageKeyParser] type. We can later
   * change the usage points to refer to [DefaultStorageKeyManager] directly, and remove this.
   */
  companion object : StorageKeyManager by DefaultStorageKeyManager
}

/** A global default thread-safe implementation of [StorageKeyManager]. */
object DefaultStorageKeyManager : StorageKeyManager {
  private val VALID_KEY_PATTERN = "^([\\w-]+)://(.*)$".toRegex()
  private var parsers = mutableMapOf<String, StorageKeyParser<*>>()

  /** Parses a raw [key] into a [StorageKey]. */
  override fun parse(rawKeyString: String): StorageKey {
    val match =
      requireNotNull(VALID_KEY_PATTERN.matchEntire(rawKeyString)) {
        "Invalid key pattern"
      }

    val protocol = match.groupValues[1]
    val contents = match.groupValues[2]
    val parser = synchronized(this) {
      requireNotNull(parsers[protocol]) {
        "No registered parsers for protocol \"$protocol\""
      }
    }

    return parser.parse(contents)
  }

  /** Registers a new [StorageKey] parser for the given [protocol]. */
  override fun addParser(parser: StorageKeyParser<*>) = synchronized(this) {
    parsers[parser.protocol] = parser
  }

  /** Resets the registered parsers to the defaults. */
  override fun reset(vararg initialSet: StorageKeyParser<*>) = synchronized(this) {
    parsers = initialSet.associateBy { it.protocol }.toMutableMap()
  }
}
